IChatCompletion Multiple Results / New Interfaces

RogerBarreto opened the pull request:
### Motivation and Context

Add potential for IChatCompletion with multiple results per prompt.

![MultiChatResultPR](https://github.com/microsoft/semantic-kernel/assets/19890735/725ecce9-5e22-4a46-a06d-b159c68807f0)

### Description

- Adds new Interfaces/Abstractions for Chat Results: `IChatResult`, `IChatStreamingResult`, `IChatMessage`
- Adds new `IChatCompletion` interface methods for Streaming and Non-Streaming results
- Moves old `IChatCompletion` interface methods `IAsyncEnumerable<string> GenerateMessageStreamAsync` and `Task<string> GenerateMessageAsync` as extensions for retro compatibility
- Changed `ChatAsText` to use the new multiple results implementations
- Makes `ChatHistory.ChatMessage` & `ChatHistory.AuthorsRole` types obsolete (discouraging usage of those implementation in favor of new `IChatMessage`
- `ChatHistory` abstraction now extends from `List<IChatMessage>`
- Updates all Samples using the new interfaces. 

### Paves the way to ... (Futures)
- Remove `ChatHistory.ChatMessage` inner class
- Remove `ChatHistory.AuthorsRole` enums 
- Make `ChatHistory` an abstract class or an `IChatHistory` interface.
- Rename `ITextCompletionResult` to `ITextResult`
- Rename `ITextCompletionStreamingResult` to `ITextStreamingResult`

### Contribution Checklist
- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

stephentoub commented:
> Moves old IChatCompletion interface methods IAsyncEnumerable<string> GenerateMessageStreamAsync and Task<string> GenerateMessageAsync as extensions for retro compatibility

What's our plan for tracking / removing such temporary measures?  Do we actually need them?

RogerBarreto commented:
> What's our plan for tracking / removing such temporary measures? Do we actually need them?

@stephentoub The plan here is to keep those as extensions while we still use it as part of the `SK internal implementations`. 

Once we moved all the `SK` implementations to use the `Result` approach those extension methods will be flagged as `Obsolete` and removed from the extensions or being kept depending on the community request.

NTaylorMullen commented on dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatCompletionStreamingResult.cs:
```diff
@@ -0,0 +1,63 @@
+        await foreach (var message in this._choice.GetMessageStreaming(cancellationToken))
+        {
+            // As it loops thru the stream, the return ChatMessage is an updated/concatenated version of the previous ChatMessage
+            chatMessage = message;
```
This keeps replacing it, not intended I assume?

NTaylorMullen commented on dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:
```diff
@@ -316,22 +297,29 @@ private static ChatCompletionsOptions CreateChatCompletionsOptions(ChatRequestSe
+    {
+        Verify.NotNullOrWhiteSpace(role);
+
+        validRole = role.ToUpperInvariant() switch
```
This is an odd way to do an `if` statement with multiple `string.Equals(...., StringComparison.OrdinalIgnoreCase)`

NTaylorMullen commented on dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatRequestSettings.cs:
```diff
@@ -42,6 +42,13 @@ public class ChatRequestSettings
+    /// Note: Because this parameter generates many completions, it can quickly consume your token quota.
+    /// Use carefully and ensure that you have reasonable settings for max_tokens and stop.
+    /// </summary>
+    public int ResultsPerPrompt { get; set; } = 1;
```
Azure SDK dropped the ball on this one. In my opinion this should just be `N`. It's what all the core APIs reference / document

RogerBarreto replied:
Not sure what you mean here, but in the docs we make no mention to any upper limit.

NTaylorMullen replied:
Ah I was referring purely to the naming. Aka `N` instead of `ResultsPerPrompt` naming

NTaylorMullen commented on dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/IChatMessage.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Microsoft.SemanticKernel.AI.ChatCompletion;
+
+public interface IChatMessage
```
As a general note for API design in .NET: Utilizing interfaces gives an "I can do" methodology where as things like abstract classes represent an "I am". In addition, once SK goes GA you'll be bound to these interfacers and if you ever want to change them you'll need to make an `IChatMessage2` in order to expand. Alternatively, if you use an abstract class you can simply add new members as virtual.

Gist: Interfaces have a place in .NET but have a lot of flaws. For instance, all of Roslyn and a lot of ASP.NET migrated away from interface representations specifically for this reason.

NTaylorMullen commented on samples/dotnet/kernel-syntax-examples/Example30_ChatWithPrompts.cs:
```diff
@@ -106,10 +107,10 @@ public static async Task RunAsync()
         // or create a new one every time, assuming that the "system message" contains all the
         // information needed.
-        var chatHistory = chatGPT.CreateNewChat(systemMessage);
+        var chatHistory = (OpenAIChatHistory)chatGPT.CreateNewChat(systemMessage);
```
Why do we need to cast to `OpenAIChatHistory`?

NTaylorMullen commented on samples/dotnet/kernel-syntax-examples/Example40_MultiChatCompletion.cs:
```diff
@@ -0,0 +1,90 @@
+            Env.Var("AZURE_OPENAI_ENDPOINT"),
+            Env.Var("AZURE_OPENAI_KEY"));
+
+        IChatCompletion completion = kernel.GetService<IChatCompletion>();
```
Ya this stinks / breaks down the kernel expectations w/ SK (shouldn't need to pull out the chat completion bit). From a high level I'd imagine services would be valuable purely for skills/connectors etc; however, once you expect an end-user to extract services from the kernel then IMO you're in rough territory / breaking boundaries between what's a service

SergeyMenshykh commented on dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatCompletionStreamingResult.cs:
```diff
@@ -0,0 +1,63 @@
+
+        if (chatMessage is null)
+        {
+            throw new AIException(AIException.ErrorCodes.UnknownError);
```
Consider adding a few details describing what happened to simplify debugging/troubleshooting process.

RogerBarreto replied:
Good point, added more details to the exception message.

SergeyMenshykh commented on dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatCompletionStreamingResult.cs:
```diff
@@ -0,0 +1,63 @@
+
+    public ChatStreamingResult(StreamingChatChoice choice)
+    {
+        this._choice = choice;
```
Should we check the parameter for null? because if it's null neither of the class methods are able to fulfill their contract.

RogerBarreto reviewed (commented):
Addressed PR Comments

shawncal commented on dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:
```diff
@@ -296,16 +270,23 @@ private static CompletionsOptions CreateCompletionsOptions(string text, Complete
-    private static ChatCompletionsOptions CreateChatCompletionsOptions(ChatRequestSettings requestSettings, ChatHistory chat)
+    private static ChatCompletionsOptions CreateChatCompletionsOptions(ChatRequestSettings requestSettings, IEnumerable<ChatMessageBase> chatHistory)
     {
+        if (requestSettings.ResultsPerPrompt < 1 ||
+            requestSettings.ResultsPerPrompt > 128)
```
tip: new tricks (optional)
```suggestion
        if (requestSettings.ResultsPerPrompt is < 1 or > 128)
```

shawncal commented on dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletion/AzureChatCompletion.cs:
```diff
@@ -83,7 +82,7 @@ public ChatHistory CreateNewChat(string? instructions = null)
         CompleteRequestSettings requestSettings,
```
Should CompleteRequestSettings have the same nullability/default rules as ChatRequestSettings?

shawncal commented on dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletion/AzureChatCompletion.cs:
```diff
@@ -54,21 +53,21 @@ public sealed class AzureChatCompletion : AzureOpenAIClientBase, IChatCompletion
         CancellationToken cancellationToken = default)
     {
-        return this.InternalGenerateChatMessageStreamAsync(chat, requestSettings ?? new(), cancellationToken);
+        return this.InternalGenerateStreamingChatMessageAsync(chat, requestSettings ?? new(), cancellationToken);
```
The public methods here have a nullable ChatRequestSettings?, but the internal implementations require this arg to be non-null (hence the new() here)... if falling back to a new is valid, I'd rather see the value of requestSettings passed through here (null or not) to the Internal* methods.

Then, only those methods, where the value is actually used, can set it to new() or otherwise ignore it when it's null. Otherwise, we get different fallback behavior across different callers (some will inject new, others will Verify.NotNull() and error out, others may make something else up).

shawncal commented on dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:
```diff
@@ -203,52 +188,41 @@ await foreach (ChatMessage message in choice.GetMessageStreaming(cancellationTok
-    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
-    /// <returns>Text generated by the remote model</returns>
-    private async Task<string> InternalCompleteTextUsingChatAsync(
+    private protected async Task<IReadOnlyList<ITextCompletionResult>> InternalGetTextCompletionAsChatAsync(
```
I think this should probably be plural `Completions`, as it returns multiple results, and to align with `GetStreamingChatCompletionsAsync`. The public methods appear plural, but internal ones are singular.

shawncal commented on dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:
```diff
@@ -203,52 +188,41 @@ await foreach (ChatMessage message in choice.GetMessageStreaming(cancellationTok
     }
 
-    private IAsyncEnumerable<string> InternalCompleteTextUsingChatStreamAsync(
+    private protected async IAsyncEnumerable<ITextCompletionStreamingResult> InternalGetTextCompletionStreamingAsChatAsync(
```
Naming: I don't thing "AsChat" is quite right here, as that implies that we're casting something to a Chat, but it's actually doing the opposite (chat data "as" text completion result).

Suggestion: Instead of calling the model in these methods, could you just create a helper that takes in an IAsyncEnum and returns the other kind?
`IAsyncEnumerable<ITextCompletionStreamingResult> AsTextCompletionStreamingResult(
      IAsyncEnumerable<IChatCompletionStreamingResult> chatStreamResult
 )`
 
Not sure if this is practical, but it would mean there is only one actual implementation hitting Chat models, and one hitting the Completion models, and then the others are just casting/converting helper methods.

shawncal commented on dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/AuthorRole.cs:
```diff
@@ -0,0 +1,95 @@
+    /// <returns> true if left and right are both null or have equivalent labels; false otherwise </returns>
+    public static bool operator ==(AuthorRole left, AuthorRole right)
+    {
+        if (Object.ReferenceEquals(left, null) != Object.ReferenceEquals(right, null))
+        {
+            return false;
+        }
+        else if (Object.ReferenceEquals(left, null))
+        {
+            return true;
+        }
```
I know this is boilerplate, and what I'm about to say matters not one bit... but in the interest of logic puzzles. Is this the same as?:
```suggestion
        if (Object.ReferenceEquals(left, null))
        {
            return Object.ReferenceEquals(right, null);
        }
        
        return left.Equals(right);
```

shawncal replied:
(note: the above diff suggestion is missing the final return in the red/removed part)...

 `else
        {
            return left.Equals(right);
        }
    }
    `

shawncal commented on dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/IChatCompletion.cs:
```diff
@@ -14,28 +14,28 @@ public interface IChatCompletion : IAIService
+    /// <param name="requestSettings">Request settings for the completion API</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>AsyncEnumerable list of different streaming chat results generated by the remote model</returns>
+    IAsyncEnumerable<IChatStreamingResult> GetStreamingChatCompletionsAsync(
```
I like the new names here!

shawncal commented on dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/IChatStreamingResult.cs:
```diff
@@ -0,0 +1,16 @@
+    /// </summary>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>Current chat message streaming content</returns>
+    IAsyncEnumerable<ChatMessageBase> GetChatMessageStreamingAsync(CancellationToken cancellationToken = default);
```
Should this not also be in the format "GetStreaming<<things>>Async"?
`GetStreamingChatMessageAsync`

shawncal reviewed (commented):
A few comments...
